### PR TITLE
test: Enhanced ibd process test

### PR DESCRIPTION
**Pull request by driftluo**
1. Enhanced ibd process test
2. Enable disconnect test

### CODE REVIEW

**Comment by keroro520** [comment]
I am sorry that I don't understand what this testing PR enhances? Could you add some more description? @driftluo

**Comment by driftluo** [comment]
> I am sorry that I don't understand what this testing PR enhances? Could you add some more description? @driftluo

I don't quite understand your question.

The old test only tested that outbound will reject the connection of the ibd node. Now also testing ibd synchronization only with outbound.

Isn't this an enhancement? Is it weakening?

**Comment by keroro520** [comment]
> The old test only tested that outbound will reject the connection of the ibd node. Now also testing ibd synchronization only with outbound.

My question is what this PR for(I didn't get from the codes).
And this sentence above is the answer.
Thanks for your reply.

File: test/src/specs/sync/ibd_process.rs [hunk 1]
```diff
@@ -25,21 +25,58 @@ impl Spec for IBDProcess {
...
+        let rpc_client6 = node6.rpc_client();
+
+        let ret = wait_until(10, || {
+            let header1 = rpc_client1.get_tip_header();
+            let header2 = rpc_client2.get_tip_header();
+            let header3 = rpc_client3.get_tip_header();
+            let header4 = rpc_client4.get_tip_header();
+            let header5 = rpc_client5.get_tip_header();
+            let header6 = rpc_client6.get_tip_header();
+            !(header1.inner.number.0 != 0
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: Double negation is hard to read.

File: test/src/specs/sync/ibd_process.rs [hunk 2]
```diff
@@ -25,21 +25,58 @@ impl Spec for IBDProcess {
...
         // will never connect
         node0.connect_uncheck(node5);
         node0.connect_uncheck(node6);
-        node0.generate_blocks(1);
 
         sleep(Duration::from_secs(5));
 
-        let rpc_client = node0.rpc_client();
+        let rpc_client0 = node0.rpc_client();
         let ret = wait_until(10, || {
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: Give `ret` a meaningful name.

File: test/src/specs/sync/ibd_process.rs [hunk 3]
```diff
@@ -25,21 +25,58 @@ impl Spec for IBDProcess {
...
+
+        // IBD only with outbound/whitelist node
+        let rpc_client1 = node1.rpc_client();
+        let rpc_client2 = node2.rpc_client();
+        let rpc_client3 = node3.rpc_client();
+        let rpc_client4 = node4.rpc_client();
+        let rpc_client5 = node5.rpc_client();
+        let rpc_client6 = node6.rpc_client();
+
+        let ret = wait_until(10, || {
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: Give `ret` a meaningful name.

File: test/src/specs/sync/ibd_process.rs [hunk 4]
```diff
@@ -25,21 +25,58 @@ impl Spec for IBDProcess {
...
+            let header6 = rpc_client6.get_tip_header();
+            !(header1.inner.number.0 != 0
+                || header1 != header6
+                || header1 != header5
+                || header1 != header4
+                || header1 != header3
+                || header1 != header2)
+        });
+
+        assert!(ret, "node 1-4 must not sync with node0");
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: But in the assertion, you are checking all the 6 nodes.

File: test/src/specs/sync/ibd_process.rs [hunk 5]
```diff
@@ -25,21 +25,58 @@ impl Spec for IBDProcess {
...
+
+        node5.connect(node0);
+        node6.connect(node0);
+
+        let ret = wait_until(10, || {
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: Give `ret` a meaningful name.

File: test/src/specs/sync/ibd_process.rs (hunk 1 above)
Comment (driftluo) [reply test/src/specs/sync/ibd_process.rs]: I think so too, but clippy tells me this way faster, I will change it back and disable clippy warning

File: test/src/specs/sync/ibd_process.rs [hunk 6]
```diff
@@ -25,21 +25,60 @@ impl Spec for IBDProcess {
...
+
+        // IBD only with outbound/whitelist node
+        let rpc_client1 = node1.rpc_client();
+        let rpc_client2 = node2.rpc_client();
+        let rpc_client3 = node3.rpc_client();
+        let rpc_client4 = node4.rpc_client();
+        let rpc_client5 = node5.rpc_client();
+        let rpc_client6 = node6.rpc_client();
+
+        #[allow(clippy::nonminimal_bool)]
```
Comment (doitian) [inline test/src/specs/sync/ibd_process.rs]: Why clippy reports this?

Comment (doitian) [reply test/src/specs/sync/ibd_process.rs]: I tried locally, it seems that we don't need this.

Comment (driftluo) [reply test/src/specs/sync/ibd_process.rs]: emmm, maybe I have an illusion. . why don't clippy report warning now?

**Comment by doitian** [comment]
bors r=doitian,jjyr,keroro520

**Pull request merged**